File: swift/Sources/CircleAI/Retail.swift
// Retail.swift
//
// Port of the Retail vertical from src/CircleAI.Retail/RetailPrimitives.cs and
// the static domain-context constants from RetailDomainContext.cs:
//   • Product, StockLevel, Sale — domain records
//   • TopSeller                 — value type for the C# tuple
//                                 `(string Sku, int Sold)` (Swift cannot Codable
//                                 a raw tuple)
//   • IRetailBoard              — products, stock, sales, revenue, top sellers
//   • InMemoryRetailBoard       — deterministic in-memory impl
//   • RetailDomainContext       — system-prompt snippet + flags
//
// The Companion-facing wrapper (RetailCompanionAdapter) is intentionally NOT
// ported.
//
// Porting notes:
//   • `decimal` → `Decimal`; `DateTimeOffset` → `Date`.
//   • `RecordSale` for an unknown SKU throws → `RetailError.unknownSku`; it also
//     decrements stock (`Stock(sku) - Quantity`).
//   • `RevenueToday` sums `UnitPrice * Quantity` over sales whose calendar day
//     (UTC) matches `now`'s day. The comparison uses a UTC calendar to mirror
//     `DateTimeOffset.Date` (which drops the time-of-day at the value's offset;
//     the C# tests use UTC timestamps, so a UTC calendar reproduces them).
//   • `TopSellersSince` groups sales at/after `since` by SKU, sums quantity,
//     orders descending by quantity, `Take(topK)`; non-positive topK throws.
//   • All state guarded by a single `NSLock`.

import Foundation

// MARK: - Records

/// A product for sale.
public struct Product: Sendable, Equatable, Codable {
    public let sku: String
    public let name: String
    public let price: Decimal
    public let currency: String
    public let category: String?

    public init(sku: String, name: String, price: Decimal, currency: String, category: String?) {
        self.sku = sku
        self.name = name
        self.price = price
        self.currency = currency
        self.category = category
    }
}

/// A stock level for a SKU.
public struct StockLevel: Sendable, Equatable, Codable {
    public let sku: String
    public let quantity: Int

    public init(sku: String, quantity: Int) {
        self.sku = sku
        self.quantity = quantity
    }
}

/// A recorded sale.
public struct Sale: Sendable, Equatable, Codable {
    public let saleId: String
    public let sku: String
    public let quantity: Int
    public let unitPrice: Decimal
    public let atUtc: Date

    public init(saleId: String, sku: String, quantity: Int, unitPrice: Decimal, atUtc: Date) {
        self.saleId = saleId
        self.sku = sku
        self.quantity = quantity
        self.unitPrice = unitPrice
        self.atUtc = atUtc
    }
}

/// A top-selling SKU with its units sold. Value type for the C# tuple
/// `(string Sku, int Sold)`.
public struct TopSeller: Sendable, Equatable, Codable {
    public let sku: String
    public let sold: Int

    public init(sku: String, sold: Int) {
        self.sku = sku
        self.sold = sold
    }
}

// MARK: - Errors

public enum RetailError: Error, Equatable, CustomStringConvertible {
    case unknownSku(String)
    case topKOutOfRange

    public var description: String {
        switch self {
        case .unknownSku(let sku): return "Unknown SKU \(sku)"
        case .topKOutOfRange: return "topK out of range"
        }
    }
}

// MARK: - Contract

/// Products, stock, sales, revenue, and top-seller analytics for the retail
/// vertical.
public protocol IRetailBoard: AnyObject, Sendable {
    func addProduct(_ p: Product)
    func getProduct(_ sku: String) -> Product?
    func setStock(_ l: StockLevel)
    func stock(_ sku: String) -> Int
    func recordSale(_ s: Sale) throws
    func revenueToday(_ now: Date) -> Decimal
    func topSellersSince(_ since: Date, topK: Int) throws -> [TopSeller]
}

public extension IRetailBoard {
    /// Overload matching the C# default `topK = 5`.
    func topSellersSince(_ since: Date) throws -> [TopSeller] {
        try topSellersSince(since, topK: 5)
    }
}

// MARK: - InMemoryRetailBoard

/// Deterministic in-memory `IRetailBoard`. All state guarded by a single
/// `NSLock`.
public final class InMemoryRetailBoard: IRetailBoard, @unchecked Sendable {
    private let lock = NSLock()
    private var products: [String: Product] = [:]
    private var stockMap: [String: Int] = [:]
    private var sales: [Sale] = []

    private static let utcCalendar: Calendar = {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "UTC")!
        return c
    }()

    public init() {}

    public func addProduct(_ p: Product) {
        lock.lock(); defer { lock.unlock() }
        products[p.sku] = p
    }

    public func getProduct(_ sku: String) -> Product? {
        lock.lock(); defer { lock.unlock() }
        return products[sku]
    }

    public func setStock(_ l: StockLevel) {
        lock.lock(); defer { lock.unlock() }
        stockMap[l.sku] = l.quantity
    }

    public func stock(_ sku: String) -> Int {
        lock.lock(); defer { lock.unlock() }
        return stockMap[sku] ?? 0
    }

    public func recordSale(_ s: Sale) throws {
        lock.lock(); defer { lock.unlock() }
        guard products[s.sku] != nil else { throw RetailError.unknownSku(s.sku) }
        sales.append(s)
        stockMap[s.sku] = (stockMap[s.sku] ?? 0) - s.quantity
    }

    public func revenueToday(_ now: Date) -> Decimal {
        lock.lock(); defer { lock.unlock() }
        let day = Self.utcCalendar.startOfDay(for: now)
        return sales
            .filter { Self.utcCalendar.startOfDay(for: $0.atUtc) == day }
            .reduce(Decimal.zero) { $0 + $1.unitPrice * Decimal($1.quantity) }
    }

    public func topSellersSince(_ since: Date, topK: Int) throws -> [TopSeller] {
        if topK <= 0 { throw RetailError.topKOutOfRange }
        lock.lock(); defer { lock.unlock() }
        var totals: [String: Int] = [:]
        for s in sales where s.atUtc >= since {
            totals[s.sku, default: 0] += s.quantity
        }
        let ordered = totals
            .map { TopSeller(sku: $0.key, sold: $0.value) }
            .sorted { $0.sold > $1.sold }
        return Array(ordered.prefix(topK))
    }
}

// MARK: - RetailDomainContext

/// Static domain-context constants for the retail vertical.
public enum RetailDomainContext {
    public static let systemPromptSnippet = "[DOMAIN: Retail] Expert retail operations assistant. Help with stock replenishment, planogram optimisation, shrinkage reduction, seasonal promotions, customer loyalty, and sales floor management. Ground advice in margin and sell-through rates. Compliance: Consumer Protection Act, POPIA."
    public static let complianceFlags: [String] = ["Consumer_Protection_Act", "POPIA", "Labour_Relations_Act"]
    public static let suggestedTools: [String] = ["pos_system", "inventory", "analytics", "promotions_engine"]
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/retail/Retail.kt
// Retail.kt
//
// Kotlin port of CircleAI.Retail (RetailPrimitives.cs +
// RetailDomainContext.cs + RetailCompanionAdapter.cs) — the C# reference
// is the EXACT spec. A deterministic in-memory retail board: products,
// stock levels, sales, and a today-revenue + top-sellers rollup.
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`; `decimal` -> `BigDecimal`;
//     `DateTimeOffset` -> `Instant`.
//   * The C# `IReadOnlyList<(string Sku, int Sold)>` named tuple is mapped to a
//     small [TopSeller] data class (Sku, Sold) so callers keep named access.
//   * `RecordSale` throws on an unknown SKU and decrements stock.
//   * `RevenueToday` sums UnitPrice*Quantity for sales on the same UTC date as
//     `now` (parity with C# `s.AtUtc.Date == now.Date`).
//   * `TopSellersSince` groups by SKU, sums quantity, orders by sold DESC, takes topK.
//   * `{value:C}` currency formatting reproduced via [fmtC] (US culture).

package com.bhengubv.circleai.retail

import com.bhengubv.circleai.companion.CompanionContext
import com.bhengubv.circleai.companion.CompanionProactiveEvent
import com.bhengubv.circleai.companion.CompanionTurn
import com.bhengubv.circleai.companion.ICompanionSession
import com.bhengubv.circleai.companion.InterfaceKind
import kotlinx.coroutines.flow.Flow
import java.math.BigDecimal
import java.text.NumberFormat
import java.time.Instant
import java.time.ZoneOffset
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Primitives (RetailPrimitives.cs)
// =====================================================================

/** A sellable product. Mirrors C# `Product`. */
data class Product(val sku: String, val name: String, val price: BigDecimal, val currency: String, val category: String?)

/** A stock level for a SKU. Mirrors C# `StockLevel`. */
data class StockLevel(val sku: String, val quantity: Int)

/** A recorded sale. Mirrors C# `Sale`. */
data class Sale(val saleId: String, val sku: String, val quantity: Int, val unitPrice: BigDecimal, val atUtc: Instant)

/** A top-seller rollup row. Mirrors the C# `(string Sku, int Sold)` tuple. */
data class TopSeller(val sku: String, val sold: Int)

/** Deterministic retail board. Mirrors C# `IRetailBoard`. */
interface IRetailBoard {
    fun addProduct(p: Product)
    fun getProduct(sku: String): Product?
    fun setStock(l: StockLevel)
    fun stock(sku: String): Int
    fun recordSale(s: Sale)
    fun revenueToday(now: Instant): BigDecimal
    fun topSellersSince(since: Instant, topK: Int = 5): List<TopSeller>
}

/** In-memory [IRetailBoard]. Mirrors C# `InMemoryRetailBoard`. */
class InMemoryRetailBoard : IRetailBoard {
    private val products = ConcurrentHashMap<String, Product>()
    private val stockLevels = ConcurrentHashMap<String, Int>()
    private val sales = mutableListOf<Sale>()
    private val lock = Any()

    override fun addProduct(p: Product) { products[p.sku] = p }
    override fun getProduct(sku: String): Product? = products[sku]

    override fun setStock(l: StockLevel) { stockLevels[l.sku] = l.quantity }
    override fun stock(sku: String): Int = stockLevels[sku] ?: 0

    override fun recordSale(s: Sale) {
        if (!products.containsKey(s.sku)) throw IllegalStateException("Unknown SKU ${s.sku}")
        synchronized(lock) {
            sales.add(s)
            stockLevels[s.sku] = stock(s.sku) - s.quantity
        }
    }

    override fun revenueToday(now: Instant): BigDecimal = synchronized(lock) {
        val today = now.atZone(ZoneOffset.UTC).toLocalDate()
        sales.filter { it.atUtc.atZone(ZoneOffset.UTC).toLocalDate() == today }
            .fold(BigDecimal.ZERO) { acc, s -> acc + s.unitPrice.multiply(BigDecimal(s.quantity)) }
    }

    override fun topSellersSince(since: Instant, topK: Int): List<TopSeller> {
        if (topK <= 0) throw IllegalArgumentException("topK must be positive")
        synchronized(lock) {
            return sales.filter { it.atUtc >= since }
                .groupBy { it.sku }
                .map { (sku, group) -> TopSeller(sku, group.sumOf { it.quantity }) }
                .sortedByDescending { it.sold }
                .take(topK)
        }
    }
}

// =====================================================================
// DomainContext (RetailDomainContext.cs)
// =====================================================================

/** Static domain context for Retail. Mirrors C# `RetailDomainContext`. */
object RetailDomainContext {
    const val SYSTEM_PROMPT_SNIPPET: String =
        "[DOMAIN: Retail] Expert retail operations assistant. Help with stock replenishment, planogram " +
            "optimisation, shrinkage reduction, seasonal promotions, customer loyalty, and sales floor " +
            "management. Ground advice in margin and sell-through rates. Compliance: Consumer Protection Act, POPIA."

    val complianceFlags: List<String> = listOf("Consumer_Protection_Act", "POPIA", "Labour_Relations_Act")

    val suggestedTools: List<String> = listOf("pos_system", "inventory", "analytics", "promotions_engine")
}

// =====================================================================
// CompanionAdapter (RetailCompanionAdapter.cs)
// =====================================================================

/** Formats like .NET `{value:C}` under the US culture. */
internal fun fmtC(value: BigDecimal): String = NumberFormat.getCurrencyInstance(Locale.US).format(value)

/** Wraps an [ICompanionSession] with the Retail snippet + helpers. Mirrors C# `RetailCompanionAdapter`. */
class RetailCompanionAdapter(private val inner: ICompanionSession) : ICompanionSession {
    override val sessionId: String get() = inner.sessionId
    override val identityId: String get() = inner.identityId
    override val interfaceKind: InterfaceKind get() = inner.interfaceKind
    override val history: List<CompanionTurn> get() = inner.history
    override val proactiveEvents: Flow<CompanionProactiveEvent> get() = inner.proactiveEvents

    override fun getContext(): CompanionContext = inner.getContext()
    override suspend fun refreshContextAsync() = inner.refreshContextAsync()
    override suspend fun signalFeedbackAsync(positive: Boolean, note: String?) =
        inner.signalFeedbackAsync(positive, note)
    override fun close() = inner.close()

    override suspend fun sendAsync(message: String): String = inner.sendAsync(enrich(message))
    override fun streamAsync(message: String): Flow<String> = inner.streamAsync(enrich(message))
    override suspend fun agentAsync(instruction: String): String = inner.agentAsync(enrich(instruction))

    private fun enrich(m: String): String = "${RetailDomainContext.SYSTEM_PROMPT_SNIPPET}\n\n$m"

    suspend fun analyseStockHealthAsync(sku: String, onHand: Int, weeklySales: Int): String =
        inner.agentAsync("Analyse stock health for SKU $sku: $onHand units on hand, $weeklySales weekly sales. Recommend reorder point, safety stock, and EOQ.")

    suspend fun planPromotionAsync(objective: String, constraints: String): String =
        inner.agentAsync("Plan a retail promotion. Objective: $objective. Constraints: $constraints. Include mechanics, discount level, marketing channels, and success metrics.")

    suspend fun optimiseProductMixAsync(topSellersJson: String, slowMoversJson: String): String =
        inner.agentAsync("Recommend product mix changes from sellers: $topSellersJson and slow: $slowMoversJson. Cover ranging, replenishment, markdown.")

    suspend fun designPromotionAsync(goal: String, category: String, budget: BigDecimal): String =
        inner.agentAsync("Design a $goal promotion for $category on $budget budget. Mechanic, channel mix, expected lift, guardrails.")

    suspend fun handleStockoutAsync(sku: String, demandSignal: String, leadTimeDays: Int): String =
        inner.agentAsync("Handle stockout of $sku (demand: $demandSignal, lead ${leadTimeDays}d). Recovery options + customer comms.")

    suspend fun reviewDailyTradingAsync(salesByCategory: String, targetRevenue: BigDecimal): String =
        inner.agentAsync("Review today's trading: $salesByCategory vs target $targetRevenue. Wins, misses, tomorrow's adjustments.")
}
